guava/src/com/google/common/io/ByteStreams.java

```diff
@@ -679,36 +691,37 @@ public static OutputStream nullOutputStream() {
    * @since 14.0 (since 1.0 as com.google.common.io.LimitInputStream)
    */
   @Beta
-  public static InputStream limit(InputStream in, long limit) {
+  public static InputStream limit(InputStream in, @NonNegative long limit) {
     return new LimitedInputStream(in, limit);
   }
 
   private static final class LimitedInputStream extends FilterInputStream {
 
-    private long left;
+    private @NonNegative long left;
     private long mark = -1;
 
-    LimitedInputStream(InputStream in, long limit) {
+    LimitedInputStream(InputStream in, @NonNegative long limit) {
       super(in);
       checkNotNull(in);
       checkArgument(limit >= 0, "limit must be non-negative");
       left = limit;
     }
 
     @Override
-    public int available() throws IOException {
+    public @NonNegative int available() throws IOException {
       return (int) Math.min(in.available(), left);
     }
 
     // it's okay to mark even if mark isn't supported, as reset won't work
     @Override
-    public synchronized void mark(int readLimit) {
+    public synchronized void mark(@NonNegative int readLimit) {
       in.mark(readLimit);
       mark = left;
     }
 
     @Override
-    public int read() throws IOException {
+    @SuppressWarnings("compound.assignment.type.incompatible") /* left can't go below 0 because it was previously checked*/
```

Comment:
I tried to reproduce the error and found the problem. `--left` issues a warning only if it is placed after the assignment `int result = in.read();`. Maybe because `read()` is not side-effect free?